text/0207-context-scoped-attributes.md

```diff
@@ -0,0 +1,278 @@
+While both are defined as resource attributes, an Azure function app hosts multiple co-deployed functions with different names and IDs.
+In PR [open-telemetry/opentelemetry-specification#2502](https://github.com/open-telemetry/opentelemetry-specification/pull/2502), this was solved by allowing to set these resource attributes on the FaaS root span instead.
+This has the drawback that the context of which function a span is executed in is lost in any child spans that may occur within the function (app).
+
+There is also issue [open-telemetry/opentelemetry-specification#1089](https://github.com/open-telemetry/opentelemetry-specification/issues/1089) “Add BeforeEnd to have a callback where the span is still writeable” which seems to be motivated by essentially the same problem, though the motivation is stated more on a technical level: “What I tried to implement is a feature called trace field in Honeycomb (reference: [AddFieldToTrace](https://godoc.org/github.com/honeycombio/beeline-go#AddFieldToTrace))”, though lately the issue seems to go into a slightly different direction.
+
+This motivation analogously also applies to other signals (metrics, logs) which will have very similar problems. E.g. the http.app attribute also looks like something you may want to split metrics by or use as a query for logs (“I want to see all logs produced by the `http.app=myapp`”).
+
+## Explanation
+
+The context-scoped attributes allows you to attach attributes to all telemetry signals emitted within a Context (or, following the usual rules of Context values, any child context thereof unless overridden). Context-scoped attributes are normal attributes, which means you can use strings, integers, floating point numbers, booleans or arrays thereof, just like for span or resource attributes. Context-scoped attributes are associated with all telemetry signals emitted while the Context containing the Context-scoped attributes is active and are available to telemetry exporters. For spans, the context within which the span is started applies. Like other telemetry APIs, Context-scoped attributes are write-only for applications. You cannot query the currently set Context-scoped attributes, they are only available on the SDK level (e.g. to telemetry exporters, Samplers and SpanProcessors).
+
+Context-scoped attributes should be thought of equivalent to adding the attribute directly to each single telemetry item it applies to.
+
+Context-scoped attributes MUST NOT be propagated cross-service, no context propagator should be implemented for them. This is because they are meant to annotate (a subset of) the spans of a single service (see also [the next section](#comp-baggage)).
+
+<a name="comp-baggage"></a>
+
+### Comparing Context-scoped attributes to Baggage
```

Comment:
I love the idea of "internal" context-attributes to pair with "external" context-attributes in Baggage.

However - I think this still needs some kind of mechanism for the interaction of Baggage + Context.  A few major points that I think limit Baggage today:

1. There is no direct interaction between Baggage and signals.  Users are *required* to do this in instrumentation.  However, this is likely a more horizontal concern.
2. The major issues with automatically attaching Baggage to signals are as follows:
  - Security: Users should be able to explicitly pick baggage the trust
  - Cardinality: Users should be able to control metric cardinality by not including baggage that would explode timeseries


As such, I think it'd be ideal if there was a mechanism where Baggage would auto-populate context-scoped attribute *upon extraction* via some kind of configuration.  You could call it a "Baggage Sampler" or a "BaggageToContextImpoter" or some such, but I think we need this horizontal interaction in the SDK.

Reply:
The goal of this OTEP was not to improve the baggage API. If this is what is desired, I think it would be best if somebody else integrates the ideas from this PR into a new OTEP that could replace this one. I don't have any experience with (or use cases for) Baggage personally, so I don't really feel up for it.